packages/@sanity/base/src/components/rovingFocus/useRovingFocus.ts

```diff
@@ -0,0 +1,153 @@
+/* eslint-disable react-hooks/exhaustive-deps */
+import {useCallback, useEffect, useState} from 'react'
+import {RovingFocusProps} from '.'
+
+const EMPTY_ARRAY = []
+
+const MUTATION_ATTRIBUTE_FILTER = ['aria-hidden', 'disabled', 'href']
+
+const FOCUSABLE =
+  'a[href], button, input, textarea, select, details, [tabindex]:not([tabindex="-1"])'
+
+function getFocusableElements(element: HTMLElement) {
+  return [...(element.querySelectorAll(FOCUSABLE) as any)].filter(
+    (el) => !el.hasAttribute('disabled') && !el.getAttribute('aria-hidden="true"')
+  ) as HTMLElement[]
+}
+
+export function useRovingFocus(props: RovingFocusProps, dependencies?: unknown[]): undefined {
+  const {direction = 'horizontal', initialFocus, loop = true, pause = false, rootElement} = props
+  const [focusedIndex, setFocusedIndex] = useState<number>(-1)
+  const [focusableElements, setFocusableElements] = useState<HTMLElement[]>([])
+
+  const deps = dependencies ? dependencies : EMPTY_ARRAY
+
+  const focusableLen = focusableElements.length
+  const lastFocusableIndex = focusableLen - 1
+
+  /**
+   * Determine what keys to listen to depending on direction
+   */
+  const nextKey = direction === 'horizontal' ? 'ArrowRight' : 'ArrowDown'
+  const prevKey = direction === 'horizontal' ? 'ArrowLeft' : 'ArrowUp'
+
+  /**
+   * Set focusable elements in state
+   */
+  const handleSetElements = useCallback(() => {
+    if (rootElement) {
+      const els = getFocusableElements(rootElement)
+
+      setFocusableElements(els)
+    }
+  }, [rootElement, ...deps])
+
+  /**
+   * Set focused index
+   */
+  const handleFocus = useCallback((index: number) => {
+    setFocusedIndex(index)
+  }, [])
+
+  /**
+   * Handle increment/decrement of focusedIndex
+   */
+  const handleKeyDown = useCallback(
+    (event) => {
+      if (pause) {
+        return
+      }
+
+      if (event.key === prevKey) {
+        event.preventDefault()
+        setFocusedIndex((prevIndex) => {
+          const next = (prevIndex + lastFocusableIndex) % focusableLen
+
+          if (!loop && next === lastFocusableIndex) {
+            return prevIndex
+          }
+
+          return next
+        })
```

Comment:
this is very tomato-tomato (so feel free to ignore) but you could use `Math.max` and `Math.min` to help handle the non-loop case e.g.

```js
if (!loop) return Math.max(0, prevIndex - 1)
```